src/transformers/modeling_funnel.py

```diff
@@ -0,0 +1,1529 @@
+            tensor = torch.cat([tensor[:, :1], t], dim=1)
+
+        ndim = tensor.ndim
+        if ndim == 2:
+            tensor = tensor[:, None, :, None]
+        elif ndim == 3:
+            tensor = tensor[:, None, :, :]
+        stride = (stride, 1)
+
+        tensor = tensor.float()
+        if mode == "mean":
+            tensor = F.avg_pool2d(tensor, stride, stride=stride, ceil_mode=True)
+        elif mode == "max":
+            tensor = F.max_pool2d(tensor, stride, stride=stride, ceil_mode=True)
+        elif mode == "min":
+            tensor = -F.max_pool2d(-tensor, stride, stride=stride, ceil_mode=True)
+        else:
+            raise NotImplementedError("The supported modes are 'mean', 'max' and 'min'.")
+
+        if ndim == 2:
+            return tensor[:, 0, :, 0]
+        elif ndim == 3:
+            return tensor[:, 0]
+        return tensor
+
+    def pre_attention_pooling(self, output, attention_struct):
+        """ Pool `output` and the proper parts of `attention_struct` before the attention layer. """
+        position_embeds, token_type_mat, attention_mask, cls_mask = attention_struct
+        if self.config.pool_q_only:
+            if self.config.attention_type == "factorized":
+                position_embeds = self.stride_pool(position_embeds[:2], 0) + position_embeds[2:]
+            token_type_mat = self.stride_pool(token_type_mat, 1)
+            cls_mask = self.stride_pool(cls_mask, 0)
+            output = self.pool_tensor(output, mode=self.config.pooling_type)
+        else:
+            self.pooling_mult *= 2
+            if self.config.attention_type == "factorized":
+                position_embeds = self.stride_pool(position_embeds, 0)
+            token_type_mat = self.stride_pool(token_type_mat, [1, 2])
+            cls_mask = self.stride_pool(cls_mask, [1, 2])
+            attention_mask = self.pool_tensor(attention_mask, mode="min")
+            output = self.pool_tensor(output, mode=self.config.pooling_type)
+        attention_struct = (position_embeds, token_type_mat, attention_mask, cls_mask)
+        return output, attention_struct
+
+    def post_attention_pooling(self, attention_struct):
+        """ Pool the proper parts of `attention_struct` after the attention layer. """
+        position_embeds, token_type_mat, attention_mask, cls_mask = attention_struct
+        if self.config.pool_q_only:
+            self.pooling_mult *= 2
+            if self.config.attention_type == "factorized":
+                position_embeds = position_embeds[:2] + self.stride_pool(position_embeds[2:], 0)
+            token_type_mat = self.stride_pool(token_type_mat, 2)
+            cls_mask = self.stride_pool(cls_mask, 1)
+            attention_mask = self.pool_tensor(attention_mask, mode="min")
+        attention_struct = (position_embeds, token_type_mat, attention_mask, cls_mask)
+        return attention_struct
+
+
+def _relative_shift_gather(x, context_len, shift):
```

Comment:
not a big fan of `x` and `y` here -> make it impossible to later rename these variables...maybe just `hidden_states` as in other models.